### PR TITLE
Add popup

### LINKED ISSUE #6: [TC][FE] Implement frontend for add popup
Story Point: 1
- [ ] create react component for pop-ups

UX mockup:
[add_trecipe_popup.pdf](https://github.com/shizuko-akamoto/Trecipe/files/4684613/add_trecipe_popup.pdf)
[add_trecipe_popup2.pdf](https://github.com/shizuko-akamoto/Trecipe/files/4703631/add_trecipe_popup2.pdf) (contains mockup for disabled create button)

### PR DESCRIPTION
**Pull request by tianyan-zhu**
- Implemented 'Add popup' component
- Added another 'Toggle Switch' component

closes #6

### CODE REVIEW

File: src/stylesheets/AddPopup.scss [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    height: 100px;
+    resize: none;
+    padding-top: 10px;
+    padding-bottom: 10px;
+}
+
+
+.button {
+    height: 40px;
+    font-size: 13px;
```
Comment (shizuko-akamoto) [inline src/stylesheets/AddPopup.scss:69]: I think Button component has a prop that you can put in custom font sizes.

File: src/stylesheets/AddPopup.scss [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    text-align: right;
```
Comment (shizuko-akamoto) [inline src/stylesheets/AddPopup.scss:70]: You can remove these right alignment related stuff if you put`justify-content: flex-end` inside `.btn` styling.

File: src/components/AddPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            <div className='AddPopup'>
+                <h1 className="Title"> New Trecipe </h1>
+                <h2> Trecipe Name *</h2>
+                <input className="input" maxLength={50} placeholder="Enter Trecipe Name" value={this.state.content} onChange={this.handleChange}/>
+                <h2> Description </h2>
+                <textarea className="input" />
+                <h2> Make Public</h2>
+                <ToggleSwitch />
+                <div className='btn'>
+                    <Button text={"Create"} disabled={!this.state.content}/>
```
Comment (shizuko-akamoto) [inline src/components/AddPopup.tsx:33]: I think `disabled={!this.state.content}` should be `disabled={_.isEmpty(this.state.content)}` ? (_.isEmpty is a utility function from lodash library we already have installed)

File: src/stylesheets/ToggleSwitch.scss [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  font-size: 2rem;
+  color: $light-gray;
+  width: 50px;
+  height: 25px;
+  background: #777777;
+  border: 1px solid $silver-gray;
+  position: relative;
+  border-radius: 13px;
+  cursor: pointer;
+  transition: border-color background-color 0.3s;
```
Comment (shizuko-akamoto) [inline src/stylesheets/ToggleSwitch.scss:18]: This part's showing red on my end? 
image

File: src/stylesheets/ToggleSwitch.scss [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  margin: 10px;
```
Comment (shizuko-akamoto) [inline src/stylesheets/ToggleSwitch.scss:19]: Could we just to top and bottom margin here? Like `margin: 10px 0;`
I think the left margin is making the toggle not align with others in the popup.
image

File: src/stylesheets/ToggleSwitch.scss (hunk 4 above)
Comment (tianyan-zhu) [reply src/stylesheets/ToggleSwitch.scss:18]: image

It is fine on my end.
This is probably the difference between vsc and webstorm

File: src/components/AddPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+
+    render() {
+        return (
+            <div className='AddPopup'>
+                <h1 className="Title"> New Trecipe </h1>
+                <h2> Trecipe Name *</h2>
+                <input className="input" maxLength={50} placeholder="Enter Trecipe Name" value={this.state.content} onChange={this.handleChange}/>
+                <h2> Description </h2>
+                <textarea className="input" />
```
[the remark this answers is not in the document]
Comment (shizuko-akamoto) [reply src/components/AddPopup.tsx:29]: I think a simple “Enter description” is enough for now?

File: src/stylesheets/ToggleSwitch.scss (hunk 4 above)
Comment (shizuko-akamoto) [reply src/stylesheets/ToggleSwitch.scss:18]: 😮

File: src/components/AddPopup.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    public readonly state: Readonly<AddPopupState> = {content: ""};
+    private createButtonRef: React.RefObject<Button>;
+    private currLength: number;
+
+    /**
+    *  Enable and Disable the 'Create Button' according to the content in Trecipe Name Input Box.
+    */
+    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        let value = e.target.value;
+        let tempLenth: number = value.length;
```
Comment (shizuko-akamoto) [inline src/components/AddPopup.tsx:36]: typo

File: src/components/AddPopup.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import "../stylesheets/AddPopup.scss"
+import  { ToggleSwitch } from "./Toggle"
+import { Button } from "./button";
+
+
+/**
+ * Add Popup state
+ * content: The content inside the Trecipe Name input box.
+ */
+export interface AddPopupState {
```
Comment (shizuko-akamoto) [inline src/components/AddPopup.tsx:12]: Are we not storing descriptions anywhere?

File: src/components/Toggle.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    public readonly state: Readonly<ToggleState> = {
+        checked: this.props.defaultChecked
+    };
+
+    /**
+    * Toggle on and off the switch.
+    */
+    toggle() {
+        this.setState({ checked: !this.state.checked })
```
Comment (shizuko-akamoto) [inline src/components/Toggle.tsx:38]: I think we need to this instead:
```
this.setState((state) => { checked: !state.checked })
```
Since we're using the previous state to update to next state. Some async stuff i don't really recall LOL.

File: src/components/button.tsx [hunk 10]
```diff
@@ -16,6 +16,7 @@ export interface ButtonProps {
     icon: IconProp;
     fontSize: number;
     onClick(e: MouseEvent<HTMLElement>): void;
+    disabled: boolean;
```
Comment (shizuko-akamoto) [inline src/components/button.tsx:19]: Can we rename this to something like `defaultDisabled`? Just to distinguish this from the state's disabled.

File: src/components/AddPopup.tsx [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    private createButtonRef: React.RefObject<Button>;
+    private currLength: number;
+
+    /**
+    *  Enable and Disable the 'Create Button' according to the content in Trecipe Name Input Box.
+    */
+    handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        let value = e.target.value;
+        let tempLenth: number = value.length;
+        if ((tempLenth > 0 && this.currLength < 1) || (tempLenth < 1 && this.currLength > 0)){
```
Comment (shizuko-akamoto) [inline src/components/AddPopup.tsx:37]: Doc: Maybe add a comment here of what this is checking?

File: src/components/Toggle.tsx (hunk 9 above)
Comment (tianyan-zhu) [reply src/components/Toggle.tsx:38]: It doesn't compile :smileyface:

File: src/index.tsx [hunk 12]
```diff
@@ -3,10 +3,11 @@ import ReactDOM from 'react-dom';
 import './index.css';
 import App from './App';
 import * as serviceWorker from './serviceWorker';
+import { AddPopup } from './components/AddPopup';
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <AddPopup />
```
Comment (tianyan-zhu) [inline src/index.tsx:10]: Will revert this change before merging.

File: src/components/Toggle.tsx (hunk 9 above)
Comment (shizuko-akamoto) [reply src/components/Toggle.tsx:38]: Probably missing a bracket or something

File: .vscode/launch.json [hunk 13]
```diff
@@ -0,0 +1,17 @@
+{
```
[the remark this answers is not in the document]
Comment (tianyan-zhu) [reply .vscode/launch.json:1]: I guess we can add another State for the description later? Maybe after they cover MongoDB.

[the remark this answers is not in the document]
Comment (tianyan-zhu) [reply .vscode/launch.json:1]: I guess this .vscode/ is just the same as .idea for IntelliJ??? some setup files
